doc/source/tune/doc_code/trial_checkpoint.py

```diff
@@ -0,0 +1,185 @@
+
+
+tuner = tune.Tuner(
+    MyTrainableClass,
+    param_space={"input_size": 64},
+    run_config=air.RunConfig(
+        stop={"training_iteration": 2},
+        checkpoint_config=air.CheckpointConfig(checkpoint_frequency=2),
+    ),
+)
+tuner.fit()
+# __class_api_checkpointing_end__
+
+# __class_api_manual_checkpointing_start__
+import random
+
+
+# to be implemented by user.
+def detect_instance_preemption():
+    choice = random.randint(1, 100)
+    # simulating a 1% chance of preemption.
+    return choice <= 1
+
+
+def train_func(self):
+    # training code
+    result = {"mean_accuracy": "my_accuracy"}
+    if detect_instance_preemption():
+        result.update(should_checkpoint=True)
+    return result
+
+
+# __class_api_manual_checkpointing_end__
+
+# __class_api_periodic_checkpointing_start__
+
+tuner = tune.Tuner(
+    MyTrainableClass,
+    run_config=air.RunConfig(
+        stop={"training_iteration": 2},
+        checkpoint_config=air.CheckpointConfig(checkpoint_frequency=10),
+    ),
+)
+tuner.fit()
+
+# __class_api_periodic_checkpointing_end__
+
+
+# __class_api_end_checkpointing_start__
+tuner = tune.Tuner(
+    MyTrainableClass,
+    run_config=air.RunConfig(
+        stop={"training_iteration": 2},
+        checkpoint_config=air.CheckpointConfig(
+            checkpoint_frequency=10, checkpoint_at_end=True
+        ),
+    ),
+)
+tuner.fit()
+
+# __class_api_end_checkpointing_end__
+
+
+# __function_api_checkpointing_start__
+from ray import tune
+from ray.air import session
+from ray.air.checkpoint import Checkpoint
+
+
+def train_func(config):
+    epochs = config.get("epochs", 2)
+    start = 0
+    loaded_checkpoint = session.get_checkpoint()
+    if loaded_checkpoint:
+        last_step = loaded_checkpoint.to_dict()["step"]
+        start = last_step + 1
+
+    for epoch in range(start, epochs):
+        # Model training here
+        # ...
+
+        # Report metrics and save a checkpoint
+        metrics = {"metric": "my_metric"}
+        checkpoint = Checkpoint.from_dict({"epoch": epoch})
+        session.report(metrics, checkpoint=checkpoint)
+
+
+tuner = tune.Tuner(train_func)
+results = tuner.fit()
+# __function_api_checkpointing_end__
+
+
+class MyModel:
+    pass
+
+
+# __function_api_checkpointing_from_dir_start__
+from ray import tune
+from ray.air import session
+from ray.air.checkpoint import Checkpoint
+
+
+# like Keras, or pytorch save methods.
+def write_model_to_dir(model, dir_path):
+    pass
+
+
+def write_epoch_to_dir(epoch: int, path: str):
+    pass
+
+
+def get_epoch_from_dir(path: str) -> int:
+    pass
+
+
+def train_func(config):
+    start = 0
+    if session.get_checkpoint() is not None:
+        loaded_checkpoint = session.get_checkpoint()
+        with loaded_checkpoint.as_directory() as loaded_checkpoint_path:
+            start = get_epoch_from_dir(loaded_checkpoint_path) + 1
+
+    for epoch in range(start, config["epochs"]):
```

Comment:
Nit: for consistency

```suggestion
    start = 1
    if session.get_checkpoint() is not None:
        loaded_checkpoint = session.get_checkpoint()
        with loaded_checkpoint.as_directory() as loaded_checkpoint_path:
            start = get_epoch_from_dir(loaded_checkpoint_path) + 1

    for epoch in range(start, config["epochs"] + 1):
```